pkg/segment/aggregations/segaggs.go

```diff
@@ -196,26 +206,247 @@ func performAggOnResult(nodeResult *structs.NodeResult, agg *structs.QueryAggreg
 	return nil
 }
 
-func performMaxRows(nodeResult *structs.NodeResult, aggs *structs.QueryAggregators, maxRows uint64, recs map[string]map[string]interface{}) error {
+// only called when headExpr has BoolExpr
+func performConditionalHeadOnHistogram(nodeResult *structs.NodeResult, headExpr *structs.HeadExpr) error {
+	fieldsInExpr := headExpr.BoolExpr.GetFields()
+	fieldToValue := make(map[string]segutils.CValueEnclosure, 0)
+
+	for _, aggregationResult := range nodeResult.Histogram {
+		newResults := make([]*structs.BucketResult, 0)
+
+		if !headExpr.Done {
+			for rowIndex, bucketResult := range aggregationResult.Results {
+				// Get the values of all the necessary fields.
+				err := getAggregationResultFieldValues(fieldToValue, fieldsInExpr, aggregationResult, rowIndex)
+				if err != nil {
+					return fmt.Errorf("performConditionalHeadOnHistogram: error while getting agg result fields values, err: %v", err)
+				}
+
+				// Evaluate the expression to a value.
+				result, err := headExpr.BoolExpr.Evaluate(fieldToValue)
+				if err != nil {
+					nullFields, err := headExpr.BoolExpr.GetNullFields(fieldToValue)
+					if err == nil && len(nullFields) > 0 {
+						// evaluation failed due to null fields
+						if headExpr.Null {
+							newResults = append(newResults, bucketResult)
+						} else if headExpr.Keeplast {
+							newResults = append(newResults, bucketResult)
+							headExpr.Done = true
+							break
+						} else {
+							headExpr.Done = true
+							break
+						}
+					} else {
+						return fmt.Errorf("performConditionalHeadOnHistogram: Error while evaluating expression on histogram, err: %v", err)
+					}
+				} else {
+					if result {
+						newResults = append(newResults, bucketResult)
+					} else {
+						// false condition so adding last record if keeplast
+						if headExpr.Keeplast {
+							newResults = append(newResults, bucketResult)
+						}
+						headExpr.Done = true
+						break
+					}
+				}
+
+				if headExpr.MaxRows > 0 && headExpr.RowsAdded == headExpr.MaxRows {
+					headExpr.Done = true
```

Comment:
Where are you incrementing the `RowsAdded`? And why is it in the else condition only? Don't you need to handle this check for `NullFields` case also?

Reply:
If there is null the evaluation will fail and give an error. So null check is only required when we have thrown an error because we cannot evaluate due to null value.